test/nestedJoin.integration.test.ts

```diff
@@ -0,0 +1,432 @@
+
+import {
+    Client,
+    createTestClient,
+    http,
+    parseUnits,
+    publicActions,
+    PublicActions,
+    TestActions,
+    WalletActions,
+    walletActions,
+} from 'viem';
+
+import {
+    Slippage,
+    NestedJoin,
+    Token,
+    replaceWrapped,
+    NestedPoolState,
+} from '../src/entities';
+import { Address } from '../src/types';
+
+import {
+    BALANCER_RELAYER,
+    BALANCER_VAULT,
+    CHAINS,
+    ChainId,
+} from '../src/utils';
+
+import {
+    approveToken,
+    findTokenBalanceSlot,
+    sendTransactionGetBalances,
+    setTokenBalance,
+} from './lib/utils/helper';
+import { authorizerAbi, vaultAbi } from '../src/abi';
+import { Relayer } from '../src/entities/relayer';
+import { NestedJoinInput } from '../src/entities/nestedJoin/types';
+
+/**
+ * Deploy the new relayer contract with the new helper address:
+ *
+ * in the mono repo:
+ * cd pkg/standalone-utils
+ * forge create --rpc-url http://0.0.0.0:8545 --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 contracts/BatchRelayerQueryLibrary.sol:BatchRelayerQueryLibrary --constructor-args "0xBA12222222228d8Ba445958a75a0704d566BF2C8"
+ *
+ * [take the address]
+ *
+ * forge create --rpc-url http://0.0.0.0:8545 --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 contracts/relayer/BalancerRelayer.sol:BalancerRelayer --constructor-args "0xBA12222222228d8Ba445958a75a0704d566BF2C8" "0xf77018c0d817da22cadbdf504c00c0d32ce1e5c2" "[paste the address]" "5"
+ *
+ * update `BALANCER_RELAYER` on constants.ts
+ *
+ */
+
+type TxInput = {
+    amountsIn: {
+        address: Address; // DAI
+        rawAmount: bigint;
+        decimals: number;
+    }[];
+    chainId: ChainId;
+    rpcUrl: string;
+    testAddress: Address;
+    nestedJoin: NestedJoin;
+    nestedPoolFromApi: NestedPoolState;
+    client: Client & PublicActions & TestActions & WalletActions;
+    useNativeAssetAsWrappedAmountIn?: boolean;
+};
+
+describe('nested join test', () => {
+    let api: MockApi;
+    let chainId: ChainId;
+    let rpcUrl: string;
+    let client: Client & PublicActions & TestActions & WalletActions;
+    let poolAddress: Address;
+    let nestedPoolFromApi: NestedPoolState;
+    let nestedJoin: NestedJoin;
+    let testAddress: Address;
+
+    beforeAll(async () => {
+        // setup mock api
+        api = new MockApi();
+
+        // setup chain and test client
+        chainId = ChainId.MAINNET;
+        rpcUrl = 'http://127.0.0.1:8545/';
+        client = createTestClient({
+            mode: 'hardhat',
+            chain: CHAINS[chainId],
+            transport: http(rpcUrl),
+        })
+            .extend(publicActions)
+            .extend(walletActions);
+
+        testAddress = (await client.getAddresses())[0];
+
+        poolAddress = '0x08775ccb6674d6bdceb0797c364c2653ed84f384'; // WETH-3POOL-BPT
+
+        // get pool state from api
+        nestedPoolFromApi = await api.getNestedPool(poolAddress);
```

Comment:
One thing in terms of UX - how does a user know if they need a nested join? 
We maybe need some helper that can show ifNested and what tokens can be used to join.

Reply:
That's an interesting question.
I think that a method to check if a pool is nested or not makes more sense at the api level.
Same as to find out which tokens could be used to join.
We could add a helper within the SDK that requests that kind info from the API and forwards that to the user.
Is that what you have in mind?

Reply:
Yeah, might be a good time to take advantage of the API. 
Maybe the user shouldn't have to think about nested vs normal join at all.
Might be possible to have a single `getPool` endpoint that will return a list of tokens that can be used to join along with the data needed to construct the call. Then the joinPool can handle which method to join with. One issue would be dealing with constraints around joining with different tokens.
```
{
  id,
  address,
  type,
  joinTokens: [] --- list of tokens that can be used to join, including nested
  poolTokens
  nested: {
    pools: [{id, type, etc}, {...}] --- data needed to construct join (same as now)
  }
}
```

Reply:
Yeah, that feels more like a business than a technical decision to me.
We seem to be leaning towards avoiding one-size-fits-all approaches elsewhere (e.g. vault, etc..), so I'd assume we would follow the same idea here.
I like a pool model that returns everything you need to perform either types of joins, but separated in a way that it's clear what's it for.
With that in mind, I'd expect something like this:
```
{
  id,
  address,
  type,
  poolTokens --- for normal join
  nested: {
    pools: [{id, type, etc}, {...}] --- data needed to construct join (same as now)
    joinTokens: [] --- list of tokens that can be used to join (nested only)
  }
}
```
Where `nested` is optional and exists only for nested pools.

Reply:
Maybe one thing that would help us answer how this should look like is: how will the FE present different options of joins on the UI?
- a single page with all join tokens (normal + nested) -> then the first approach makes more sense.
- two separate pages/tabs with each type of join and it's inputs -> then the second approach makes more sense.

Reply:
Thats a good question. I'll make a note to bring this up with Daniel in the standup on Monday.

Reply:
Registering over here that we decided to keep both regular and nested joins on separate flows, which means that the second approach, with `joinTokens` within `nested` object makes more sense.
Since this format will be used for `exits` as well, should we rename `joinTokens` by `mainTokens`?

Reply:
Thanks Bruno. Yeah, lets keep consitency.